File: android/src/main/java/com/duysolo/speechrecognition/SpeechRecognitionOptions.kt
package com.duysolo.speechrecognition

import android.media.AudioFormat
import android.speech.RecognizerIntent
import com.facebook.react.bridge.ReadableMap

class SpeechRecognitionOptions(
    val interimResults: Boolean = false,
    val lang: String = "en-US",
    val continuous: Boolean = false,
    val maxAlternatives: Int = 5,
    val contextualStrings: List<String>? = null,
    val requiresOnDeviceRecognition: Boolean = false,
    val addsPunctuation: Boolean = false,
    val androidIntentOptions: Map<String, Any?>? = null,
    val androidRecognitionServicePackage: String? = null,
    val audioSource: AudioSourceOptions? = null,
    val recordingOptions: RecordingOptions? = null,
    val androidIntent: String? = RecognizerIntent.ACTION_RECOGNIZE_SPEECH,
    val iosTaskHint: String? = null,
    val iosCategory: Map<String, Any>? = null,
    val volumeChangeEventOptions: VolumeChangeEventOptions? = null,
    val iosVoiceProcessingEnabled: Boolean = false,
) {
    companion object {
        fun fromReadableMap(map: ReadableMap): SpeechRecognitionOptions {
            val contextualStrings = if (map.hasKey("contextualStrings")) {
                val arr = map.getArray("contextualStrings")
                (0 until (arr?.size() ?: 0)).mapNotNull { arr?.getString(it) }
            } else null

            val androidIntentOptions = if (map.hasKey("androidIntentOptions")) {
                map.getMap("androidIntentOptions")?.toHashMap()
            } else null

            val audioSource = if (map.hasKey("audioSource")) {
                map.getMap("audioSource")?.let { AudioSourceOptions.fromReadableMap(it) }
            } else null

            val recordingOptions = if (map.hasKey("recordingOptions")) {
                map.getMap("recordingOptions")?.let { RecordingOptions.fromReadableMap(it) }
            } else null

            val volumeChangeEventOptions = if (map.hasKey("volumeChangeEventOptions")) {
                map.getMap("volumeChangeEventOptions")?.let { VolumeChangeEventOptions.fromReadableMap(it) }
            } else null

            return SpeechRecognitionOptions(
                interimResults = if (map.hasKey("interimResults")) map.getBoolean("interimResults") else false,
                lang = if (map.hasKey("lang")) map.getString("lang") ?: "en-US" else "en-US",
                continuous = if (map.hasKey("continuous")) map.getBoolean("continuous") else false,
                maxAlternatives = if (map.hasKey("maxAlternatives")) map.getInt("maxAlternatives") else 5,
                contextualStrings = contextualStrings,
                requiresOnDeviceRecognition = if (map.hasKey("requiresOnDeviceRecognition")) map.getBoolean("requiresOnDeviceRecognition") else false,
                addsPunctuation = if (map.hasKey("addsPunctuation")) map.getBoolean("addsPunctuation") else false,
                androidIntentOptions = androidIntentOptions,
                androidRecognitionServicePackage = if (map.hasKey("androidRecognitionServicePackage")) map.getString("androidRecognitionServicePackage") else null,
                audioSource = audioSource,
                recordingOptions = recordingOptions,
                androidIntent = if (map.hasKey("androidIntent")) map.getString("androidIntent") else RecognizerIntent.ACTION_RECOGNIZE_SPEECH,
                iosTaskHint = if (map.hasKey("iosTaskHint")) map.getString("iosTaskHint") else null,
                volumeChangeEventOptions = volumeChangeEventOptions,
                iosVoiceProcessingEnabled = if (map.hasKey("iosVoiceProcessingEnabled")) map.getBoolean("iosVoiceProcessingEnabled") else false,
            )
        }
    }
}

class VolumeChangeEventOptions(
    val enabled: Boolean = false,
    val intervalMillis: Int? = null,
) {
    companion object {
        fun fromReadableMap(map: ReadableMap): VolumeChangeEventOptions {
            return VolumeChangeEventOptions(
                enabled = if (map.hasKey("enabled")) map.getBoolean("enabled") else false,
                intervalMillis = if (map.hasKey("intervalMillis")) map.getInt("intervalMillis") else null,
            )
        }
    }
}

class RecordingOptions(
    val persist: Boolean = false,
    val outputDirectory: String? = null,
    val outputFileName: String? = null,
    val outputSampleRate: Int? = null,
    val outputEncoding: String? = null,
) {
    companion object {
        fun fromReadableMap(map: ReadableMap): RecordingOptions {
            return RecordingOptions(
                persist = if (map.hasKey("persist")) map.getBoolean("persist") else false,
                outputDirectory = if (map.hasKey("outputDirectory")) map.getString("outputDirectory") else null,
                outputFileName = if (map.hasKey("outputFileName")) map.getString("outputFileName") else null,
                outputSampleRate = if (map.hasKey("outputSampleRate")) map.getInt("outputSampleRate") else null,
                outputEncoding = if (map.hasKey("outputEncoding")) map.getString("outputEncoding") else null,
            )
        }
    }
}

class AudioSourceOptions(
    val uri: String = "",
    val audioEncoding: Int = AudioFormat.ENCODING_PCM_16BIT,
    val sampleRate: Int = 16000,
    val audioChannels: Int = 1,
    val chunkDelayMillis: Long? = null,
) {
    companion object {
        fun fromReadableMap(map: ReadableMap): AudioSourceOptions {
            return AudioSourceOptions(
                uri = if (map.hasKey("uri")) map.getString("uri") ?: "" else "",
                audioEncoding = if (map.hasKey("audioEncoding")) map.getInt("audioEncoding") else AudioFormat.ENCODING_PCM_16BIT,
                sampleRate = if (map.hasKey("sampleRate")) map.getInt("sampleRate") else 16000,
                audioChannels = if (map.hasKey("audioChannels")) map.getInt("audioChannels") else 1,
                chunkDelayMillis = if (map.hasKey("chunkDelayMillis")) map.getInt("chunkDelayMillis").toLong() else null,
            )
        }
    }
}

class GetSupportedLocaleOptions(
    val androidRecognitionServicePackage: String? = null,
)

class TriggerOfflineModelDownloadOptions(
    val locale: String = "en-US",
)

File: ios/SpeechRecognitionOptions.swift
import Speech

struct SpeechRecognitionOptions {
  var interimResults: Bool = false
  var lang: String = "en-US"
  var continuous: Bool = false
  var maxAlternatives: Int = 5
  var contextualStrings: [String]? = nil
  var requiresOnDeviceRecognition: Bool = false
  var addsPunctuation: Bool = false
  var recordingOptions: RecordingOptions? = nil
  var audioSource: AudioSourceOptions? = nil
  var iosTaskHint: IOSTaskHint? = nil
  var iosCategory: SetCategoryOptions? = nil
  var volumeChangeEventOptions: VolumeChangeEventOptions? = nil
  var iosVoiceProcessingEnabled: Bool? = false

  init(from dict: NSDictionary) {
    self.interimResults = dict["interimResults"] as? Bool ?? false
    self.lang = dict["lang"] as? String ?? "en-US"
    self.continuous = dict["continuous"] as? Bool ?? false
    self.maxAlternatives = dict["maxAlternatives"] as? Int ?? 5
    self.contextualStrings = dict["contextualStrings"] as? [String]
    self.requiresOnDeviceRecognition = dict["requiresOnDeviceRecognition"] as? Bool ?? false
    self.addsPunctuation = dict["addsPunctuation"] as? Bool ?? false
    self.iosVoiceProcessingEnabled = dict["iosVoiceProcessingEnabled"] as? Bool ?? false

    if let recordDict = dict["recordingOptions"] as? NSDictionary {
      self.recordingOptions = RecordingOptions(from: recordDict)
    }
    if let audioDict = dict["audioSource"] as? NSDictionary {
      self.audioSource = AudioSourceOptions(from: audioDict)
    }
    if let hint = dict["iosTaskHint"] as? String {
      self.iosTaskHint = IOSTaskHint(rawValue: hint)
    }
    if let catDict = dict["iosCategory"] as? NSDictionary {
      self.iosCategory = SetCategoryOptions(from: catDict)
    }
    if let volDict = dict["volumeChangeEventOptions"] as? NSDictionary {
      self.volumeChangeEventOptions = VolumeChangeEventOptions(from: volDict)
    }
  }
}

struct VolumeChangeEventOptions {
  var enabled: Bool? = false
  var intervalMillis: Int? = nil

  init(from dict: NSDictionary) {
    self.enabled = dict["enabled"] as? Bool ?? false
    self.intervalMillis = dict["intervalMillis"] as? Int
  }
}

enum IOSTaskHint: String {
  case unspecified
  case dictation
  case search
  case confirmation

  var sfSpeechRecognitionTaskHint: SFSpeechRecognitionTaskHint {
    switch self {
    case .unspecified: return .unspecified
    case .dictation: return .dictation
    case .search: return .search
    case .confirmation: return .confirmation
    }
  }
}

struct RecordingOptions {
  var persist: Bool = false
  var outputDirectory: String? = nil
  var outputFileName: String? = nil
  var outputSampleRate: Double? = nil
  var outputEncoding: String? = nil

  init(from dict: NSDictionary) {
    self.persist = dict["persist"] as? Bool ?? false
    self.outputDirectory = dict["outputDirectory"] as? String
    self.outputFileName = dict["outputFileName"] as? String
    self.outputSampleRate = dict["outputSampleRate"] as? Double
    self.outputEncoding = dict["outputEncoding"] as? String
  }
}

struct AudioSourceOptions {
  var uri: String = ""
  var audioEncoding: Int? = nil
  var sampleRate: Int? = 16000
  var audioChannels: Int? = 1
  var chunkDelayMillis: Int? = nil

  init(from dict: NSDictionary) {
    self.uri = dict["uri"] as? String ?? ""
    self.audioEncoding = dict["audioEncoding"] as? Int
    self.sampleRate = dict["sampleRate"] as? Int ?? 16000
    self.audioChannels = dict["audioChannels"] as? Int ?? 1
    self.chunkDelayMillis = dict["chunkDelayMillis"] as? Int
  }
}

enum CategoryParam: String {
  case ambient, soloAmbient, playback, record, playAndRecord, multiRoute

  var avCategory: AVAudioSession.Category {
    switch self {
    case .ambient: return .ambient
    case .soloAmbient: return .soloAmbient
    case .playback: return .playback
    case .record: return .record
    case .playAndRecord: return .playAndRecord
    case .multiRoute: return .multiRoute
    }
  }
}

enum CategoryOptionsParam: String {
  case mixWithOthers, duckOthers, interruptSpokenAudioAndMixWithOthers
  case allowBluetooth, allowBluetoothA2DP, allowAirPlay
  case defaultToSpeaker, overrideMutedMicrophoneInterruption

  var avCategoryOption: AVAudioSession.CategoryOptions {
    switch self {
    case .mixWithOthers: return .mixWithOthers
    case .duckOthers: return .duckOthers
    case .interruptSpokenAudioAndMixWithOthers: return .interruptSpokenAudioAndMixWithOthers
    case .allowBluetooth: return .allowBluetooth
    case .allowBluetoothA2DP: return .allowBluetoothA2DP
    case .allowAirPlay: return .allowAirPlay
    case .defaultToSpeaker: return .defaultToSpeaker
    case .overrideMutedMicrophoneInterruption:
      if #available(iOS 14.5, *) {
        return .overrideMutedMicrophoneInterruption
      } else {
        return .mixWithOthers
      }
    }
  }
}

enum ModeParam: String {
  case `default`, gameChat, measurement, moviePlayback, spokenAudio
  case videoChat, videoRecording, voiceChat, voicePrompt

  var avMode: AVAudioSession.Mode {
    switch self {
    case .default: return .default
    case .gameChat: return .gameChat
    case .measurement: return .measurement
    case .moviePlayback: return .moviePlayback
    case .spokenAudio: return .spokenAudio
    case .videoChat: return .videoChat
    case .videoRecording: return .videoRecording
    case .voiceChat: return .voiceChat
    case .voicePrompt: return .voicePrompt
    }
  }
}

struct SetCategoryOptions {
  var category: CategoryParam = .playAndRecord
  var categoryOptions: [CategoryOptionsParam] = [.duckOthers]
  var mode: ModeParam = .measurement

  init(from dict: NSDictionary) {
    if let cat = dict["category"] as? String, let c = CategoryParam(rawValue: cat) {
      self.category = c
    }
    if let opts = dict["categoryOptions"] as? [String] {
      self.categoryOptions = opts.compactMap { CategoryOptionsParam(rawValue: $0) }
    }
    if let m = dict["mode"] as? String, let mode = ModeParam(rawValue: m) {
      self.mode = mode
    }
  }

  // Default init
  init() {}
}

struct SetAudioSessionActiveOptions {
  var notifyOthersOnDeactivation: Bool? = true
}
